sdk/core/azure-core-amqp/inc/azure/core/amqp/common/async_operation_queue.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <azure/core/context.hpp>
+#include <condition_variable>
+#include <list>
+#include <mutex>
+#include <thread>
+#include <tuple>
+
+namespace Azure { namespace Core { namespace Amqp { namespace Common { namespace _internal {
+
+  /** An AsyncOperationQueue represents a queue of "results" for an async operation.
+   *
+   * It expresses a relatively simple API contract. The code which produces results calls
+   * "CompleteOperation" which sets the result, and a consumer calls WaitForResult which reads from
+   * the AsyncOperationQueue. WaitForResult will block until a result is available.
+   */
+  template <typename... T> class AsyncOperationQueue {
+
+  public:
+    AsyncOperationQueue() = default;
+    ~AsyncOperationQueue() = default;
+
+    void CompleteOperation(T... operationParameters)
+    {
+      std::unique_lock<std::mutex> lock(m_operationComplete);
+      m_operationQueue.push_back(std::make_unique<std::tuple<T...>>(
+          std::make_tuple<T...>(std::forward<T>(operationParameters)...)));
+      lock.unlock();
+      m_operationCondition.notify_one();
+    }
+
+    template <class... Poller>
+    std::unique_ptr<std::tuple<T...>> WaitForPolledResult(
+        Azure::Core::Context context,
+        Poller&... pollers)
+    {
+      do
+      {
+        {
+          std::unique_lock<std::mutex> lock(m_operationComplete);
+          if (!m_operationQueue.empty())
+          {
+            std::unique_ptr<std::tuple<T...>> rv;
+            rv = std::move(m_operationQueue.front());
+            m_operationQueue.pop_front();
+            return rv;
+          }
+          if (context.IsCancelled())
+          {
+            return nullptr;
+          }
+        }
+        std::this_thread::yield();
+
+        // Note: We need to call Poll() *outside* the lock because the poller is going to call the
+        // CompleteOperation function.
+        Poll(pollers...);
```

Comment:
The name `Poll` makes this seem like an LRO concept. Should we rename it to something else? We can't really use `Operation<T>` here because that's tied to sending a request and getting back a response.

https://github.com/Azure/azure-sdk-for-cpp/blob/93434329d19cfa0dbec24cbedd1d0d9b7c33af64/sdk/core/azure-core/inc/azure/core/operation.hpp#L170-L175

Reply:
The construct used in uAMQP/azure-c-shared-utility is "work" (as in `DoWork`). I dislike that construct because it isn't actually correct - what is happening in the `DoWork` function is that the caller is functionally polling the construct (connection or transport typically) to see if forward progress can be made, then moving the internal state along based on whether there is something to do - normally that function is called "polling", which is why I named it as `Poll`.

If you have a better term for "See if there is something to do, and if there is, do it", I'm more than willing to consider it.

But the verb "Poll" is extremely common, and just because it's used in LROs doesn't mean it can't be used elsewhere for similar purposes.

Reply:
Given it is in the internal namespace, it seems reasonable to keep using it.